README.md
# stweam

[![build status](https://img.shields.io/travis/tanem/stweam/master.svg?style=flat-square)](https://travis-ci.org/tanem/stweam)
[![coverage status](https://img.shields.io/coveralls/tanem/stweam.svg?style=flat-square)](https://coveralls.io/r/tanem/stweam)
[![npm version](https://img.shields.io/npm/v/stweam.svg?style=flat-square)](https://www.npmjs.com/package/stweam)
[![npm downloads](https://img.shields.io/npm/dm/stweam.svg?style=flat-square)](https://www.npmjs.com/package/stweam)
[![dependency status](https://david-dm.org/tanem/stweam.svg?style=flat-square)](https://david-dm.org/tanem/stweam)
[![devDependency status](https://david-dm.org/tanem/stweam/dev-status.svg?style=flat-square)](https://david-dm.org/tanem/stweam#info=devDependencies)

Streams the public Twitter stream.

Note that the output is the "raw" Twitter stream, so you'll need to [process it appropriately](https://dev.twitter.com/docs/streaming-apis/processing). Processing was deliberately left out of this module, so that you can choose your poison(s) via npm.

## Requirements

 * Node.js version 0.11.x (for the `harmony` flag which exposes generators)
 * Keys obtained from dev.twitter.com after [setting up a new app](https://apps.twitter.com/app/new)

## Installation

```
$ npm install stweam --save
```

## Example

```js
// => example.js

var Stweam = require('stweam');
var stream = require('stream');
var dest = new stream.PassThrough();

var stweam = new Stweam({
  consumerKey: 'consumerKey',
  consumerSecret: 'consumerSecret',
  token: 'token',
  tokenSecret: 'tokenSecret'
});

// Optionally hook into log messages...
stweam.on('info', function(msg){
  // Do something with msg.
});

// ... or warn messages.
stweam.on('warn', function(msg){
  // Do something with msg.
});

// stweam is a Stream.
stweam.pipe(dest);

stweam
  .language('fr')
  .track('beaker')
  .follow('12345')
  .start();
```

To run:

```
$ node --harmony example.js
```

Also, check out [chirp](https://github.com/tanem/chirp) to see `Stweam` in action.

## API

### var stweam = new Stweam(opts)

Initialise a new `Stweam` with the given `opts`.

`opts` should contain:

 * consumerKey
 * consumerSecret
 * token
 * tokenSecret

### stweam.track(keywords)

Set the phrases that will determine what is delivered on the stream.

See: [track](https://dev.twitter.com/docs/streaming-apis/parameters#track).

### stweam.language(language)

Set the Tweet language, defaults to `en`.

See: [language](https://dev.twitter.com/docs/streaming-apis/parameters#language).

### stweam.follow(userids)

Set the users whose Tweets will be delivered on the stream.

See: [follow](https://dev.twitter.com/docs/streaming-apis/parameters#follow).

### stweam.start()

Start the app.

## Events

Note: I like [Bunyan's opinions](https://github.com/trentm/node-bunyan#levels) on log levels :)

### stweam.on('info', function(msg){})

Provides detail about regular operation, rather than spamming stdout.

### stweam.on('warn', function(msg){})

Might want to take a closer look at this one.

## Tests

To run:

```
$ npm run test:spec
```

To generate a coverage report:

```
$ npm run test:cov
```

## Credits

 * [tweets](https://github.com/benfoxall/tweets)
 * [urlgrey](https://github.com/cainus/urlgrey)


CHANGELOG.md
## v1.0.15 / 2016-02-19

 * ([37410b4](https://github.com/tanem/stweam/commit/37410b4c76ca8b915529430cb5bb3f89a6fdc01d)) Add dependency badges (closes [#12](https://github.com/tanem/stweam/issues/12))

## v1.0.14 / 2016-02-06

 * ([d2ebc58](https://github.com/tanem/stweam/commit/d2ebc58bc55866ad77eddde87a5741234190b169)) Update license (closes [#11](https://github.com/tanem/stweam/issues/11))

## v1.0.13 / 2015-11-07

 * ([9bb5870](https://github.com/tanem/stweam/commit/9bb5870e5454aa3eec7df68e4de136ad59031b89)) Use npm run scripts instead of Make (fixes [#1](https://github.com/tanem/stweam/issues/1))

## v1.0.12 / 2015-11-05

 * ([e378b18](https://github.com/tanem/stweam/commit/e378b181599dd19d33834d64e7c2764053b8717b)) Update README badges (fixes [#10](https://github.com/tanem/stweam/issues/10))

## 1.0.11 / 2015-10-29

 * ([ea7d453](https://github.com/tanem/stweam/commit/ea7d45391445705417d5fe8cf1ea170174d732d9)) Add .editorconfig (fixes [#7](https://github.com/tanem/stweam/issues/7))

## v1.0.10 / 2015-10-26

 * ([a626a4b](https://github.com/tanem/stweam/commit/a626a4b8b34ba862974bd4a4e6816a594742102b)) Switch from jshint to eslint (fixes [#4](https://github.com/tanem/stweam/issues/4))

## v1.0.9 / 2015-10-25

 * ([f8dc8f9](https://github.com/tanem/stweam/commit/f8dc8f98d64be3983526ce3cbc7f7b48fed95865)) Update LICENSE (fixes [#6](https://github.com/tanem/stweam/issues/6))

## v1.0.8 / 2015-08-16

 * ([e22ca18](https://github.com/tanem/stweam/commit/e22ca18387bd5ba13627ad70726e6e83a5c5c3b3)) Tweak Makefile

## v1.0.7 / 2015-03-11

 * ([02300cf](https://github.com/tanem/stweam/commit/02300cf9f8ab610888a9646506161f174cd4e6ec)) Update LICENSE

## v1.0.6 / 2014-12-04

 * ([4d7672a](https://github.com/tanem/stweam/commit/4d7672ab311a42c70f296ac91fbb111ad43500cc)) Ocd

## v1.0.5 / 2014-11-18

 * ([7a547d8](https://github.com/tanem/stweam/commit/7a547d82bbcfef4a68a51c4135e2de86a9afc2c0)) Update README

## v1.0.4 / 2014-11-08

 * ([18164a6](https://github.com/tanem/stweam/commit/18164a6db489a08bedb7003278d39f2ff6cf7a25)) Update README

## v1.0.3 / 2014-10-21

 * ([1c1e163](https://github.com/tanem/stweam/commit/1c1e16312c15f3421645802fb57b084a14f8f8d9)) Tidy up readme
 * ([466b3e5](https://github.com/tanem/stweam/commit/466b3e5f215a353310e7181211054c7203822c5c)) Add commit links to CHANGELOG

## v1.0.1 / 2014-09-30

 * ([e4bd6f7](https://github.com/tanem/stweam/commit/e4bd6f7462755f68fece8cb0700d1ef65d3da1f2)) Generate changelog using mkclog

## v1.0.0 / 2014-09-12

 * ([1b831d9](https://github.com/tanem/stweam/commit/1b831d938eb43918b0b73167c93ac8cfb5c9d1d2)) v1.0.0

## v0.8.6 / 2014-08-24

 * ([612171d](https://github.com/tanem/stweam/commit/612171d3474bd1079c73315d790a532bd6d24bf3)) Update README

## v0.8.5 / 2014-08-24

 * ([1c6f0c2](https://github.com/tanem/stweam/commit/1c6f0c26406583cf4688b0025881c7cf501b3612)) Add follow method
 * ([f2df723](https://github.com/tanem/stweam/commit/f2df723fc45ac7be3b706b5caeec1ab935390088)) Add missing changelog entry
 * ([820c638](https://github.com/tanem/stweam/commit/820c63867bfad9f8158031ebdd054046a9f2c09c)) Add missing changelog entry

## v0.8.4 / 2014-08-16

 * ([eebdc1f](https://github.com/tanem/stweam/commit/eebdc1fba0cfa66e8161d224f35f54ffdbc3d029)) Add warn log level

## v0.8.3 / 2014-08-14

 * ([36ef48a](https://github.com/tanem/stweam/commit/36ef48af63d7605df4b88bccdb31d6f857b94bea)) Add chirp example to README

## v0.8.2 / 2014-07-20

 * ([06b37a1](https://github.com/tanem/stweam/commit/06b37a1aa5c8414738867fddea12f3d2d29b7df8)) Add http status code to log messages

## v0.8.1 / 2014-07-16

 * ([776f556](https://github.com/tanem/stweam/commit/776f556f341dc27a9a24b07baee2ab123a97d459)) Update README

## v0.8.0 / 2014-07-16

 * ([8273dba](https://github.com/tanem/stweam/commit/8273dbac5cace9efc469a64b5e990ad507dcb561)) Remove built in parsing

## v0.7.0 / 2014-07-12

 * ([d269921](https://github.com/tanem/stweam/commit/d269921eb34d4ce2bc7769f7c21d33f37f53f0ea)) Remove object mode streaming

## v0.6.0 / 2014-07-06

 * ([d21c348](https://github.com/tanem/stweam/commit/d21c348d5582954f0e00b9ca2b3c5e8407245ac3)) Update README

## v0.5.0 / 2014-07-06

 * ([953de57](https://github.com/tanem/stweam/commit/953de57dedc015081e66b2b49845506517ef3c08)) Add language method

## v0.4.0 / 2014-07-06

 * ([7274e28](https://github.com/tanem/stweam/commit/7274e286a6cc8dfb36cf51cdfb55dfe67a2558ad)) Push the entire tweet object if the receive array is empty

## v0.3.2 / 2014-07-05

 * ([2f06d56](https://github.com/tanem/stweam/commit/2f06d56e15a4a38ee3fadc62fc2737d18812b140)) Update README

## v0.3.1 / 2014-07-05

 * ([dbdea66](https://github.com/tanem/stweam/commit/dbdea668f8eebca410ecb28a55a97e76ef678d3a)) Update description

## v0.3.0 / 2014-07-05

 * ([28dd1e7](https://github.com/tanem/stweam/commit/28dd1e7846ab099bcc3e8bae9b4e1bd1a07ef6b7)) Turn Stweam into a Transform stream

## v0.2.4 / 2014-07-03

 * ([d6f6c81](https://github.com/tanem/stweam/commit/d6f6c81c4ff9e0b8fb2f85f3e54b77db8d8cf1b7)) Update docs
 * ([fda20e9](https://github.com/tanem/stweam/commit/fda20e9d32f0f596bd08f573eed59a92d051e946)) Tweak Makefile

## v0.2.3 / 2014-06-29

 * ([e0053c1](https://github.com/tanem/stweam/commit/e0053c1a05b323144cc8d867f416e2440b1f8f7e)) Add .npmignore
 * ([bb9c9cb](https://github.com/tanem/stweam/commit/bb9c9cbdfd727a8be9d65dbebc145381a507439c)) Add .npmignore
 * ([dc6748a](https://github.com/tanem/stweam/commit/dc6748ae87d4fcc21f341ba41d7416fdb8d60a4e)) Document default language

## v0.2.2 / 2014-06-28

 * ([86c8607](https://github.com/tanem/stweam/commit/86c8607cb8459575b29957a56e04039e3e18ae97)) Update rate limited backoff strategy name
 * ([d5645ca](https://github.com/tanem/stweam/commit/d5645ca6e8449873186d14adbe97e70701be0253)) Add info events
 * ([cfd5b90](https://github.com/tanem/stweam/commit/cfd5b90f85bce4a490e2a242e00b571d418005cd)) Stream tweet text directly to Stweam

## v0.2.1 / 2014-06-28


test/parser.test.js
var expect = require('expect.js');
var Parser = require('../lib/parser');
var stream = require('stream');

describe('Parser', function(){

  var parser;

  beforeEach(function(){
    parser = new Parser();
  });

  it('should parse the buffer and push the tweet string', function(done){
    var source = stream.PassThrough();
    source.pipe(parser);
    source.write(new Buffer('{ "text"'));
    source.write(new Buffer(': "bar'));
    source.end(new Buffer('" }\r\n{ "text'));

    parser
      .on('data', function(chunk){
        expect(chunk.toString()).to.eql('{ "text": "bar" }');
      })
      .on('end', done);
  });

});

test/stweam.test.js
var rewire = require('rewire');
var expect = require('expect.js');
var sinon = require('sinon');
var events = require('events');
var stream = require('stream');

describe('Stweam', function(){

  var Stweam;
  var stweam;
  var clock;

  // Fake timers need to be set up first so they're picked up by rewire.
  beforeEach(function(){
    clock = sinon.useFakeTimers();
    Stweam = rewire('../lib/stweam');
    stweam = new Stweam({
      consumerKey: 'key',
      consumerSecret: 'secret',
      token: 'token',
      tokenSecret: 'secret'
    });
  });

  afterEach(function(){
    clock.restore();
  });

  describe('construction', function(){

    it('should throw an error if consumerKey is not set', function(){
      expect(function(){
        new Stweam();
      }).to.throwException(/consumer key is required/);
    });

    it('should throw an error if consumerSecret is not set', function(){
      expect(function(){
        new Stweam({
          consumerKey: 'key'
        });
      }).to.throwException(/consumer secret is required/);
    });

    it('should throw an error if token is not set', function(){
      expect(function(){
        new Stweam({
          consumerKey: 'key',
          consumerSecret: 'secret'
        });
      }).to.throwException(/token is required/);
    });

    it('should throw an error if tokenSecret is not set', function(){
      expect(function(){
        new Stweam({
          consumerKey: 'key',
          consumerSecret: 'secret',
          token: 'token'
        });
      }).to.throwException(/token secret is required/);
    });

  });

  describe('backoff strategies', function(){

    var backoff;

    beforeEach(function(){
      backoff = { linear: sinon.stub(), exponential: sinon.stub() };
      Stweam.__set__('backoff', backoff);
      new Stweam({
        consumerKey: 'key',
        consumerSecret: 'secret',
        token: 'token',
        tokenSecret: 'secret'
      });
    });

    it('should include a linear network backoff strategy that starts at 0 and maxes at 16s', function(){
      expect(backoff.linear.args[0]).to.eql([0, 16000]);
    });

    it('should include an exponential http error backoff strategy that starts at 5s and maxes at 320s', function(){
      expect(backoff.exponential.args[0]).to.eql([5000, 320000]);
    }); 

    it('should include an exponential rate limited backoff strategy that starts at 60s and has no maximum', function(){
      expect(backoff.exponential.args[1]).to.eql([60000]);
    });

  });

  describe('Twitter request', function(){

    it('should POST with the correct parameters', function(){
      var postStub = sinon.stub();
      Stweam.__set__('request', { post: postStub });
      stweam._keywords = 'surfing';
      stweam._follow = '1234,5678';

      stweam._getRequest();

      expect(postStub.args[0][0]).to.eql({
        url: 'https://stream.twitter.com/1.1/statuses/filter.json',
        oauth: {
          consumer_key: 'key',
          consumer_secret: 'secret',
          token: 'token',
          token_secret: 'secret'
        },
        form: {
          track: 'surfing',
          language: 'en',
          follow: '1234,5678'
        }
      });

    });

  });

  describe('prior to connecting', function(){

    it('should clear an existing stall timeout', function(){
      var clearTimeoutStub = sinon.stub();
      Stweam.__set__({
        clearTimeout: clearTimeoutStub
      });
      stweam.stallTimeout = {};
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();

      expect(clearTimeoutStub.args[0][0]).to.be(stweam.stallTimeout);
    });

    it('should abort an existing request', function(){
      var requestStub = { abort: sinon.stub() };
      stweam.request = requestStub;
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();

      expect(requestStub.abort.calledOnce).to.be(true);
    });

  });

  describe('upon receiving a response', function(){
    
    it('should attempt reconnection after 90s if no data is received', function(){
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      var connectStub = sinon.stub(stweam, '_connect');
      stweam.request.emit('response', response);
      clock.tick(1000 * 90);

      expect(connectStub.calledOnce).to.be(true);
      clock.restore();
    });

    it('should cancel the stall timeout if data is received', function(){
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      var connectStub = sinon.stub(stweam, '_connect');
      stweam.request.emit('response', response);
      response.emit('data', new Buffer(''));
      clock.tick(1000 * 90);

      expect(connectStub.calledOnce).to.be(false);
      clock.restore();
    });

  });

  describe('upon receiving a request error', function(){

    it('should attempt reconnects according to the network backoff strategy', function(){
      var networkErrorBackoffStub = sinon.stub(
        stweam,
        'networkErrorBackoff',
        function(cb){
          cb();
        }
      );
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      var connectStub = sinon.stub(stweam, '_connect');
      stweam.request.emit('error');

      expect(networkErrorBackoffStub.calledOnce).to.be(true);
      expect(connectStub.calledOnce).to.be(true);
    });

  });

  describe('upon receiving a 420 response code', function(){

    it('should attempt reconnects according to the rate limited backoff strategy', function(){
      var rateLimitedErrorBackoffStub = sinon.stub(
        stweam,
        'rateLimitedErrorBackoff',
        function(cb){
          cb();
        }
      );
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      var connectStub = sinon.stub(stweam, '_connect');
      response.statusCode = 420;
      stweam.request.emit('response', response);

      expect(rateLimitedErrorBackoffStub.calledOnce).to.be(true);
      expect(connectStub.calledOnce).to.be(true);
    });

  });

  describe('upon receiving a response code >= 500', function(){

    it('should attempt reconnects according to the rate limited backoff strategy', function(){
      var rateLimitedErrorBackoffStub = sinon.stub(
        stweam,
        'rateLimitedErrorBackoff',
        function(cb){
          cb();
        }
      );
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      var connectStub = sinon.stub(stweam, '_connect');
      response.statusCode = 500;
      stweam.request.emit('response', response);

      expect(rateLimitedErrorBackoffStub.calledOnce).to.be(true);
      expect(connectStub.calledOnce).to.be(true);
    });

  });

  describe('upon receiving a response code > 200, < 500 and != 420', function(){

    it('should attempt reconnects according to the http error backoff strategy', function(){
      var httpErrorBackoffStub = sinon.stub(
        stweam,
        'httpErrorBackoff',
        function(cb){
          cb();
        }
      );
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      var connectStub = sinon.stub(stweam, '_connect');
      response.statusCode = 401;
      stweam.request.emit('response', response);

      expect(httpErrorBackoffStub.calledOnce).to.be(true);
      expect(connectStub.calledOnce).to.be(true);
    });

  });

  describe('upon receiving a valid response code', function(){
  
    it('should reset the backoff strategies', function(){
      var backoffStub = sinon.stub(stweam, '_initBackoffs');
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      stweam.request.emit('response', response);

      expect(backoffStub.calledOnce).to.be(true);
    });

    it('should pipe the twitter response through itself', function(done){
      var response = new stream.PassThrough();
      sinon.stub(stweam, '_getRequest').returns(new events.EventEmitter());

      stweam.start();
      stweam.request.emit('response', response);

      stweam
        .on('data', function(chunk){
          expect(chunk.toString()).to.be('{ "foo": "bar" }\r\n');
        })
        .on('end', done);

      response.end(new Buffer('{ "foo": "bar" }\r\n'));
    });

    // We know the piping is set up correctly from the above test,
    // so just checking `push` is called correctly.
    it('should push the tweet', function(){
      var pushStub = sinon.stub(stweam, 'push');
      stweam._transform('{ "foo": "bar", "baz": "qux" }', null, function(){});
      expect(pushStub.args[0][0]).to.eql('{ "foo": "bar", "baz": "qux" }');
    });

  });
  
  describe('track method', function(){

    it('should update the keywords property', function(){
      stweam.track('foo');
      expect(stweam._keywords).to.be('foo');
    });

  });

  describe('language method', function(){

    it('should update the language property', function(){
      stweam.language('foo');
      expect(stweam._language).to.eql('foo');
    });

  });

  describe('follow method', function(){

    it('should update the follow property', function(){
      stweam.follow('12345');
      expect(stweam._follow).to.be('12345');
    });

  });

  describe('start method', function(){

    it('should start the app', function(){
      var connectStub = sinon.stub(stweam, '_connect');
      stweam.start();
      expect(connectStub.calledOnce).to.be(true);
    });

  });

});

test/backoff.test.js
var rewire = require('rewire');
var expect = require('expect.js');
var co = require('co');

var backoff = rewire('../lib/backoff');

// Assuming `co-sleep` will function correctly, so this thunk mock just
// returns immediately.
backoff.__set__('sleep', function(){
  return function(cb){
    cb();
  };
});

describe('Backoff', function(){

  describe('linear strategy', function(){

    it('should handle starting values', function(){
      co(function*(){
        var timeout = yield backoff.linear(1000);
        expect(timeout).to.be(1250);
      })();
    });

    it('should increment the sleep value by 250ms each time', function(){
      var linear = backoff.linear(0);
      co(function*(){
        var timeout = yield linear;
        expect(timeout).to.be(250);
        timeout = yield linear;
        expect(timeout).to.be(500);
        timeout = yield linear;
        expect(timeout).to.be(750);
      })();
    });

    it('should not increment the sleep value once max is reached', function(){
      var linear = backoff.linear(0, 500);
      co(function*(){
        var timeout = yield linear;
        expect(timeout).to.be(250);
        timeout = yield linear;
        expect(timeout).to.be(500);
        timeout = yield linear;
        expect(timeout).to.be(500);
      })();
    });

  });

  describe('exponential strategy', function(){

    it('should handle starting values', function(){
      co(function*(){
        var timeout = yield backoff.exponential(1000);
        expect(timeout).to.be(2000);
      })();
    });

    it('should double the sleep value each time', function(){
      var exponential = backoff.exponential(100);
      co(function*(){
        var timeout = yield exponential;
        expect(timeout).to.be(200);
        timeout = yield exponential;
        expect(timeout).to.be(400);
        timeout = yield exponential;
        expect(timeout).to.be(800);
      })();
    });

    it('should not increment the sleep value once max is reached', function(){
      var exponential = backoff.exponential(100, 400);
      co(function*(){
        var timeout = yield exponential;
        expect(timeout).to.be(200);
        timeout = yield exponential;
        expect(timeout).to.be(400);
        timeout = yield exponential;
        expect(timeout).to.be(400);
      })();
    });

  });

});

lib/parser.js
var stream = require('stream');

/**
 * Expose `Parser`.
 */

module.exports = Parser;

/**
 * Initialise a new `Parser`.
 *
 * @api public
 */

function Parser() {
  stream.Transform.call(this);
  this.delimiter = '\r\n';
  this.buffer = '';
}

Parser.prototype = Object.create(stream.Transform.prototype);

/**
 * Adds chunks to the buffer. Once it hits the delimiter
 * it'll push the tweet string.
 *
 * @param {Buffer|String} chunk
 * @param {String} encoding
 * @param {Function} done
 * @api private
 */

Parser.prototype._transform = function (chunk, encoding, done) {
  var str = chunk.toString();
  var delimiterIndex = str.indexOf(this.delimiter);
  if (~delimiterIndex) {
    this.buffer += str.slice(0, delimiterIndex);
    this.push(this.buffer);
    this.buffer = str.slice(delimiterIndex + 2); 
  } else {
    this.buffer += str;
  }
  done();
};

lib/backoff.js
var sleep = require('co-sleep');

/**
 * Returns a generator that sleeps for `timeout` ms,
 * increasing `timeout` by 250ms each time it is called.
 *
 * @param {Number} start
 * @param {Number} max
 * @return {Function} generator
 * @api public
 */

exports.linear = function(start, max){
  var timeout = start;
  return function*(){
    yield sleep(timeout);
    if (max && timeout >= max) return timeout;
    return timeout += 250;
  };
};

/**
 * Returns a generator that sleeps for `timeout` ms,
 * doubling `timeout` each time it is called.
 *
 * @param {Number} start
 * @param {Number} max
 * @return {Function} generator
 * @api public
 */

exports.exponential = function(start, max){
  var timeout = start;
  return function*(){
    yield sleep(timeout);
    if (max && timeout >= max) return timeout;
    return timeout *= 2;
  };
};

lib/stweam.js
var request = require('request');
var stream = require('stream');
var co = require('co');
var backoff = require('./backoff');

/**
 * Expose `Stweam`.
 */

module.exports = Stweam;

/**
 * Initialise a new `Stweam` with the given `opts`.
 *
 * @param {String} opts.consumerKey
 * @param {String} opts.consumerSecret
 * @param {String} opts.token
 * @param {String} opts.tokenSecret
 * @api public
 */

function Stweam(opts) {
  stream.Transform.call(this);
  opts = opts || {};
  if (!opts.consumerKey) throw new Error('consumer key is required');
  if (!opts.consumerSecret) throw new Error('consumer secret is required');
  if (!opts.token) throw new Error('token is required');
  if (!opts.tokenSecret) throw new Error('token secret is required');
  this.consumerKey = opts.consumerKey;
  this.consumerSecret = opts.consumerSecret;
  this.token = opts.token;
  this.tokenSecret = opts.tokenSecret;
  this._keywords = '';
  this._language = 'en';
  this._follow = '';
  this._initBackoffs();
}

Stweam.prototype = Object.create(stream.Transform.prototype);

/**
 * Initialise the backoff timers.
 *
 * @see https://dev.twitter.com/docs/streaming-apis/connecting#Test_backoff_strategies
 * @api private
 */

Stweam.prototype._initBackoffs = function(){
  this.networkErrorBackoff = backoff.linear(0, 16 * 1000);
  this.httpErrorBackoff = backoff.exponential(5 * 1000, 320 * 1000);
  this.rateLimitedErrorBackoff = backoff.exponential(60 * 1000);
};

/**
 * Connect to Twitter's public statuses stream.
 *
 * @see https://dev.twitter.com/docs/api/1.1/post/statuses/filter
 * @return {Request}
 * @api private
 */

Stweam.prototype._getRequest = function(){
  return request.post({
    url: 'https://stream.twitter.com/1.1/statuses/filter.json',
    oauth: {
      consumer_key: this.consumerKey,
      consumer_secret: this.consumerSecret,
      token: this.token,
      token_secret: this.tokenSecret
    },
    form: {
      follow: this._follow,
      track: this._keywords,
      language: this._language
    }
  });
};

/**
 * Pushes the tweet.
 *
 * @param {Buffer|String} chunk
 * @param {String} encoding
 * @param {Function} done
 * @api private
 */

Stweam.prototype._transform = function(chunk, encoding, done){
  this.push(chunk);
  done();
};  

/**
 * Handle the connection plus it's associated events.
 *
 * @api private
 */

Stweam.prototype._connect = function(){

  if (this.stallTimeout) clearTimeout(this.stallTimeout);
  if (this.request) this.request.abort();

  this.request = this._getRequest();

  this.request.on('error', function(){
    this.emit('warn', 'request error -> reconnecting with network error backoff strategy');
    co(function*(){
      yield this.networkErrorBackoff;
      this._connect();
    }).call(this);
  }.bind(this));

  this.request.on('response', function(response){

    // https://dev.twitter.com/docs/streaming-apis/connecting#Stalls
    this.stallTimeout = setTimeout(function(){
      this.emit('warn', 'response stalled -> reconnecting');
      this._connect();
    }.bind(this), 90 * 1000);

    response.on('data', function(){
      clearTimeout(this.stallTimeout);
    }.bind(this));

    var code = response.statusCode;

    // Rate limited backoff.
    // https://dev.twitter.com/discussions/20479
    if (code === 420 || code >= 500) {
      this.emit('warn', 'rate limited response [' + code + '] -> reconnecting with rate limited backoff strategy');
      co(function*(){
        yield this.rateLimitedErrorBackoff;
        this._connect();
      }).call(this);
      return;
    }
    
    // Other HTTP error backoff.    
    if (code > 200) {
      this.emit('warn', 'http error response [' + code + '] -> reconnecting with http error backoff strategy');
      co(function*(){
        yield this.httpErrorBackoff;
        this._connect();
      }).call(this);
      return;
    }

    this.emit('info', 'valid response received');

    // We have a valid response so can reset the backoffs.
    this._initBackoffs();
    response.pipe(this);

  }.bind(this));

};

/**
 * Set the phrases that will determine what is delivered on the stream.
 *
 * @see https://dev.twitter.com/docs/streaming-apis/parameters#track
 * @param {String} keywords
 * @return {Stweam} self
 * @api public
 */

Stweam.prototype.track = function(keywords){
  this._keywords = keywords;
  return this;
};

/**
 * Set the Tweet language.
 *
 * @see https://dev.twitter.com/docs/streaming-apis/parameters#language
 * @param {String} language
 * @return {Stweam} self
 * @api public
 */

Stweam.prototype.language = function(language){
  this._language = language;
  return this;
};

/**
 * Set the users whose Tweets will be delivered on the stream.
 *
 * @see https://dev.twitter.com/docs/streaming-apis/parameters#follow
 * @param {String} userids
 * @return {Stweam} self
 * @api public
 */

Stweam.prototype.follow = function(userids){
  this._follow = userids;
  return this;
};

/**
 * Start the app.
 *
 * @api public
 */

Stweam.prototype.start = function(){
  this._connect();
};
